Add size control logic and test.

markscheel opened the pull request:
## Proposed changes

Adds the logic and control parameter computation for size control.

This is done via an abstract base class SizeControlState
that has derived classes for each concrete state of the size
control system.

The logic (and control errors) are added in such a way that the
code can be tested independently.  There is (intentionally) no
dependence on Strahlkorpers, FunctionsOfTime, ZeroCrossingPredictors,
or the control system framework itself.

The thing that calls the SizeControlState member functions
(and thus needs to know about Strahlkorpers, FunctionsOfTime, etc so
that it can compute the arguments to those member functions)
will be added in a later PR.

*Note:*  The logic currently lacks some of the size control states that are present in SpEC.  Those states will be added in a later PR.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Add size control logic and test.

This is done via an abstract base class SizeControlState
that has derived classes for each concrete state of the size
control system.

The logic (and control errors) are added in such a way that the
code can be tested independently.  There is (intentionally) no
dependence on Strahlkorpers, FunctionsOfTime, ZeroCrossingPredictors,
or the control system framework itself.

The thing that calls the SizeControlState member functions
(and thus needs to know about Strahlkorpers, FunctionsOfTime, etc so
that it can compute the arguments to those member functions)
will be added in a later PR.

knelli2 commented on src/ControlSystem/ControlErrors/CMakeLists.txt:
```diff
@@ -1,11 +1,37 @@
   HEADERS
   Expansion.hpp
   Rotation.hpp
+  SizeControlInfo.hpp
+  SizeControlState.hpp
+  SizeControlStateAhSpeed.hpp
+  SizeControlStateDeltaR.hpp
+  SizeControlStateInitial.hpp
```
If you don't have too much future code that depends on this current PR, could you just add a new directory `SizeControl` or `Size` inside `ControlSystem/ControlErrors` and remove the "SizeControl" prefix from all the files? I think it would look a lot cleaner if we didn't have a bunch of files prefixed with the same thing. I'm getting SpEC flashbacks from this 😅

nilsdeppe replied:
I agree, having a `Size` directory would be good. I also would like to avoid `Control` in all the names since everything here is in the control system. E.g. having directories `Size` and `Errors` to try and keep names somewhat shorter.

knelli2 replied:
Once this PR is merged, I can change the `ControlErrors` directory and namepsace to `Errors` to be consistent with everything else.

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlState.hpp:
```diff
@@ -0,0 +1,95 @@
+/// Each state has its own logic (the 'update' function) that
+/// determines values of certain parameters (i.e. the things in
+/// SizeControlInfo), including whether the control system should
+/// transition to a different state.
+class SizeControlState {
```
All of these classes you've added are in the global namespace. Again, if you don't have too much future code that depends on this PR, could you move all these classes into a new namespace `SizeControl` or `Size` that's within the `control_system::ControlErrors` namespace? (also removing the "SizeControl" prefix from all the classes)

knelli2 reviewed (changes requested):
I haven't looked at the internals of everything yet, only the naming structure.

If you decide to do these changes, you can squash directly.

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlInfo.hpp:
```diff
@@ -0,0 +1,62 @@
+/// Holds information that is saved between calls of SizeControl.
+struct SizeControlInfo {
+  SizeControlLabel state{SizeControlLabel::Initial};
+  // The current damping time associated with size control.
```
Make these doxygen comments, throughout the PR.

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlInfo.cpp:
```diff
@@ -0,0 +1,22 @@
+      t_comoving_char_speed(comoving_char_speed_crossing_time),
+      t_delta_radius(delta_radius_crossing_time) {
+  if (t_char_speed > 0.0) {
+    if (t_delta_radius > 0.0 and t_delta_radius <= t_char_speed) {
```
[Optional] This sort of thing would be simpler if you didn't use 0 to mean infinity.  The same pattern occurs throughout this PR.

markscheel replied:
I think I want to wait to see what the final ZeroCrossingPredictor looks like (because t_char_speed etc will come from the ZeroCrossingPredictor) before changing this from zero to something else.  But yes, instead of zero I am open to using a `std::optional` or maybe `std:::numeric_limits<double>::infinity` to denote "doesn't cross zero or cannot determine when it crosses zero".

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlState.hpp:
```diff
@@ -0,0 +1,95 @@
+/// transition to a different state.
+class SizeControlState {
+ public:
+  virtual ~SizeControlState() {};
```
`= default`

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+  }();
+
+  const bool comoving_decreasing_slower_than_char_speeds =
+      not(crossing_time_info.t_char_speed > 0.0 &&
```
`and`

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+  if (char_speed_is_in_danger) {
+    constexpr double min_char_speed_increase_factor = 1.01;
+    if (info->target_char_speed < min_char_speed) {
+      // We are already in state 1, and we are in danger.
```
"state 1" -> ?

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+      // to continue to push CharSpeedMin downward.  Instead, increase
+      // TargetSpeed to be above CharSpeedMin.
+      // We don't do this if DeltaRexIsInDanger, because for DeltaRex
+      // in danger, we might need to drive CharSpeed lower.
```
Update the variable names in this comment.

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlInfo.hpp:
```diff
@@ -0,0 +1,62 @@
+///   to SizeControlInfo::target_char_speed.
+/// - DeltaR: drives the minimum distance between the horizon and the excision
+///   boundary to be constant in time.
+enum class SizeControlLabel { Initial, AhSpeed, DeltaR };
```
I don't know exactly how this will be used, but would it make sense to store the actual state object instead of a label?

markscheel replied:
Storing the actual state object might work if we used a std::variant or a unique_ptr to the base class so that the things stored would all have the same type, but then each of the classes would need a virtual function that returns an ID telling what actual type it is.  Unless I'm missing something, that seems more complicated than an enum class.

wthrowe replied:
I was thinking a unique_ptr, yes.

It's hard for me to comment on what implementation makes sense when I don't know what you are going to do with this.  I don't know why  you want a type ID, for example.  Do you have something I can look at, even if it's still a work in progress?

markscheel replied:
I think a unique_ptr would work.  I think the only time that I would need a type ID is for diagnostic output that would indicate what the current state is.

If a unique_ptr is stored in SizeControlInfo (which is serializable now), then presumably all the derived State classes need pup functions (which would be empty since the classes store no data)?

markscheel replied:
Oh, I also use a type ID for the test (to check that the states are correct).  But without a type ID, I can check if `dynamic_cast` of the unique_ptr to a ptr of the correct type succeeds.

wthrowe replied:
You wouldn't need pup functions (those can be inherited), but you would need some macros and PUP_ID stuff.  Kind of annoying and maybe not worth the trouble.  I'll leave which way to go up to you.

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlInfo.hpp:
```diff
@@ -0,0 +1,62 @@
+enum class SizeControlLabel { Initial, AhSpeed, DeltaR };
+
+/// Holds information that is saved between calls of SizeControl.
+struct SizeControlInfo {
```
Do these structs need to be serializable?

knelli2 replied:
Forgot to add this when I did my review, but yes. If these structs are going to be held inside a `control_system::protocols::ControlError` (which it seems like they will be), then they will need to be serializable.

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlInfo.hpp:
```diff
@@ -0,0 +1,62 @@
+  // t_char_speed is the time (relative to the current time) when the
+  // minimum comoving characteristic speed is predicted to cross zero
+  // (or zero if the minimum comoving characteristic speed is increasing).
+  double t_comoving_char_speed;
```
Update comment to match variable name. (Same for `t_delta_radius`)

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlState.hpp:
```diff
@@ -0,0 +1,95 @@
+        min_comoving_char_speed(min_comoving_speed),
+        control_error_delta_r(control_err_delta_r) {}
+  // min_char_speed is the minimum over the excision boundary
+  // of Eq. 89 of ArXiv:1211.6079.
```
When promoting to doxygen comment, add a citation link for all of these.

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlState.hpp:
```diff
@@ -0,0 +1,95 @@
+                             const double control_err_delta_r)
+      : min_char_speed(min_speed),
+        min_comoving_char_speed(min_comoving_speed),
+        control_error_delta_r(control_err_delta_r) {}
```
Is the constructor actually needed here? I believe you can omit the constructor and have
```cpp
SizeControlStateUpdateArgs update_args{0.1, 0.2, 0.3};
```
will work just fine. (Similar to how `LinkedMessageId<T>` is used) Same for `ControlSignalArgs` below. However, if you plan to do other things in the constructor in the future then it's alright to leave it.

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlState.hpp:
```diff
@@ -0,0 +1,95 @@
+/// the Y00 coefficient of the shape map so that the minimum
+/// characteristic speed is driven towards a target value, and state
+/// SizeControlLabel::DeltaR controls the Y00 coefficient of the shape
+/// map so that the minimum difference between the horizon radius and
```
While technically true, I believe we are going to have separate functions of time and separate maps for size and shape. So the `Y00` component of the actual shape map will always be zero because we will have a size map which acts as the `Y00` component.

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlState.hpp:
```diff
@@ -0,0 +1,95 @@
+  virtual ~SizeControlState() {};
+  /// Updates the SizeControlInfo in `info`.  Notice that `info`
+  /// includes a state, which might be different than the current
+  /// state upon return (if so, then the caller will update the
+  /// current state).
```
Could you move this comment out of parentheses and change it to something like
```
upon return. It is the caller's responsibility to check if the current state has changed.
```
I think this is an important enough implementation detail that it shouldn't just be a note at the end.

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+  // Note that delta_radius_is_in_danger and char_speed_is_in_danger
+  // can be different for different SizeControlStates.
+  //
+  // The value of 20 causes the control system to panic easily if
+  // delta_radius is decreasing quickly.
```
Could you give a little more info on this number? Maybe how you came to choose it (even if it was just experimentation in SpEC)? That way in a year when I'm looking back at this I don't just see a magic `20.0` that says "this is the right value".

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+    }
+    info->suggested_time_scale = crossing_time_info.t_char_speed;
+  } else if (delta_radius_is_in_danger) {
+    constexpr double target_speed_decrease_factor = 0.125;
+    constexpr double time_tolerance_for_delta_r =
+        0.25 * time_tolerance_for_delta_r_in_danger;
```
Why these values? (Again, experimentation in SpEC is valid) Same for the rest of the `constexpr` values in this PR that aren't self explanatory (like `Y00`)

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+    } else {
+      info->discontinuous_change_has_occurred = true;
+      info->state = SizeControlLabel::DeltaR;
+      info->suggested_time_scale = crossing_time_info.t_delta_radius;
+      // TODO: PreferState3OverState2 goes here.
```
If `State3` is one of those states you'll be adding in a future PR, could you change `State3` and `State2` to more descriptive names? (Same throughout this PR).

knelli2 commented on src/ControlSystem/ControlErrors/SizeControlStateAhSpeed.cpp:
```diff
@@ -0,0 +1,112 @@
+
+double AhSpeed::control_signal(
+    const SizeControlInfo& info,
+    const SizeControlStateControlSignalArgs& control_signal_args) const {
+  const double Y00 = sqrt(0.25 / M_PI);
```
`constexpr`?

markscheel replied:
constexpr didn't work; it gave an error about sqrt not being contexpr.

markscheel replied:
https://stackoverflow.com/questions/8622256/in-c11-is-sqrt-defined-as-constexpr

knelli2 reviewed (changes requested):
Ok now this is a full review.

markscheel commented:
Made requested changes (except for removing Labels, which I am still thinking about, and changing zero="There is no crossing time", which I want to wait for ZeroCrossingPredictor).

Moved files around and put into a new namespace.  Added a pup function for SizeControlInfo (I don't think the other structs need to be serializable because they are created and used locally and never saved)

knelli2 commented on tests/Unit/ControlSystem/ControlErrors/Test_SizeControlStates.cpp:
```diff
@@ -14,11 +14,11 @@
-    const SizeControlState& size_control_state,
-    const SizeControlStateUpdateArgs& update_args,
-    const CrossingTimeInfo& crossing_time_info) {
+    const gsl::not_null<control_system::size::Info*> info,
```
[Optional] Put the whole test in the `control_system::size` namespace so you don't have to keep retyping it? I'm fine either way.

knelli2 reviewed (changes requested):
Looks good!

wthrowe commented on src/ControlSystem/ControlErrors/SizeControlInfo.hpp:
```diff
@@ -12,26 +14,75 @@
+    } else {
+      auto state_as_size_t = static_cast<size_t>(state);
+      p | state_as_size_t;
+    }
```
I think you can just `p | state;` for an enum.

markscheel replied:
the enum goes away with my next commit.

markscheel commented:
Added a new commit that implements @wthrowe suggestion of ditching labels and storing only objects.
This makes many things more complicated (i.e. you need pup stuff and clone functions), but the part that gets simplified is that now you don't need to manually build a correspondence between labels and objects.

knelli2 reviewed (commented):
All changes look good to me. Squash after @wthrowe gives his review.

wthrowe commented on src/ControlSystem/ControlErrors/Size/AhSpeed.hpp:
```diff
@@ -3,18 +3,27 @@
+
+  WRAPPED_PUPable_decl_template(AhSpeed); // NOLINT
+  explicit AhSpeed(CkMigrateMessage* const /*msg*/) {}
+  void pup(PUP::er& /*p*/) override {}
```
I believe the base class `pup` has to be called.  The easiest way to do that is to just not declare a `pup` and let the function be inherited.  Same for the other classes.

wthrowe commented on src/ControlSystem/ControlErrors/Size/AhSpeed.hpp:
```diff
@@ -3,18 +3,27 @@
 namespace control_system::size::States {
 class AhSpeed : public State {
  public:
+  AhSpeed() : State() {}
```
`= default;`  Same for the other classes.

wthrowe commented on src/ControlSystem/ControlErrors/Size/State.hpp:
```diff
@@ -67,9 +76,48 @@ struct ControlSignalArgs {
+class State : public PUP::able {
  public:
-  virtual ~State() = default;
+  State(){};
```
`= default;`

wthrowe reviewed (changes requested):
You can squash these in directly.

markscheel commented:
Made changes and squashed everything.

markscheel commented:
CodeCov told me that I wasn't testing serialization of control_system::size::Info, so I added that test and squashed.
